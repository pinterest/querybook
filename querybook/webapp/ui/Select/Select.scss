.Select {
    display: inline-block;
    position: relative;

    select {
        -webkit-appearance: none;
        background-color: inherit;
        color: inherit;
        border-radius: var(--border-radius);
        border: var(--border);
        font-size: inherit;
        padding: 8px 32px 8px 8px;

        &:hover {
            border-color: var(--hover-border-color);
        }
        &:active,
        &:focus {
            border-color: var(--focus-border-color);
        }
    }

    &::after {
        z-index: 4;
        display: block;
        position: absolute;
        top: 38%;
        right: 12px;
        height: 8px;
        width: 8px;
        border: 2px solid transparent;
        border-radius: 2px;
        border-right: 0;
        border-top: 0;
        border-color: inherit;
        transform: rotate(-45deg);
        transform-origin: center;
        pointer-events: none;
        content: ' ';
    }
    &.Select-transparent {
        select {
            font-size: var(--small-text-size);
            border: none;
            padding: 4px 32px 4px 4px;
        }
        &::after {
            top: 32%;
        }
    }
    &.Select-full-width {
        width: 100%;
        select {
            width: 100%;
        }
    }

    &.small {
        select {
            font-size: var(--xxsmall-text-size);
        }
    }
}
