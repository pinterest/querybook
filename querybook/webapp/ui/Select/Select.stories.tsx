import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { Select, ISelectProps, makeSelectOptions } from './Select';

export const _Select = (args: ISelectProps) => {
    const [value, setValue] = React.useState('1');

    const onChange = (newVal) => {
        setValue(newVal.target.value);
    };

    return (
        <Select {...args} value={value} onChange={onChange}>
            {makeSelectOptions([
                { value: 'option one', key: '1' },
                { value: 'option two', key: '2' },
            ])}
        </Select>
    );
};

_Select.args = {
    fullWidth: false,
    transparent: false,

    withDeselect: false,
    disabled: false,
};

export default {
    title: 'Form/Select',
    decorators: [centered],
};
