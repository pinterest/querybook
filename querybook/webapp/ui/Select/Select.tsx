import React from 'react';
import clsx from 'clsx';
import './Select.scss';

export const DESELECT_VALUE = '';
export interface ISelectProps {
    name?: string;
    disabled?: boolean;
    value: any;
    onChange: (evt: React.ChangeEvent<HTMLSelectElement>) => any;
    className?: string;
    fullWidth?: boolean;
    transparent?: boolean;

    withDeselect?: boolean;
}

export const Select: React.FunctionComponent<ISelectProps> = ({
    name,
    disabled,
    value,
    onChange,
    className = '',
    children,

    fullWidth,
    transparent,

    withDeselect,
}) => {
    const deselectOption = withDeselect ? (
        <option value={DESELECT_VALUE} key="">
            {'Choose one'}
        </option>
    ) : null;

    const selectDOM = (
        <select
            name={name}
            disabled={disabled || false}
            value={value ?? DESELECT_VALUE}
            onChange={onChange}
        >
            {deselectOption}
            {children}
        </select>
    );

    const wrapperClassName = clsx({
        Select: true,
        'Select-transparent': transparent,
        'Select-full-width': fullWidth,
        [className]: Boolean(className),
    });

    const controlClassName = clsx({
        control: true,
    });

    return (
        <div className={controlClassName}>
            <div className={wrapperClassName}>{selectDOM}</div>
        </div>
    );
};

export type IOptions = Array<
    | {
          key: any;
          value: string;
      }
    | string
>;

export function makeSelectOptions(options: IOptions) {
    return options.map(
        // option can be a direct string or an object with key and value property
        (option) =>
            typeof option === 'string' ? (
                <option value={option} key={option}>
                    {option}
                </option>
            ) : (
                <option value={option.key} key={option.key}>
                    {option.value}
                </option>
            )
    );
}
