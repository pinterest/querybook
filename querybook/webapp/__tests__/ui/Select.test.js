import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Select, makeSelectOptions } from '../../ui/Select/Select';

it('renders without crashing', () => {
    shallow(<Select value="test" onChange={() => null} />);
});

describe('Select behavior', () => {
    it('selects the correct field', () => {
        const onChangeMock = jest.fn();
        const wrapper = mount(
            <Select
                value="test"
                onChange={(event) => onChangeMock(event.target.value)}
            >
                {makeSelectOptions(['test', 'test1', 'test2'])}
            </Select>
        );
        wrapper.find('select').instance().selectedIndex = 2;
        wrapper.find('select').simulate('change');

        expect(wrapper.find('option')).toHaveLength(3);
        expect(onChangeMock).toHaveBeenCalledTimes(1);
        expect(onChangeMock).toHaveBeenCalledWith('test2');
    });

    it('deselects correctly', () => {
        const onChangeMock = jest.fn();
        const wrapper = mount(
            <Select
                value="test"
                onChange={(event) => onChangeMock(event.target.value)}
                withDeselect
            >
                {makeSelectOptions(['test', 'test1', 'test2'])}
            </Select>
        );
        wrapper.find('select').instance().selectedIndex = 0;
        wrapper.find('select').simulate('change');

        expect(wrapper.find('option')).toHaveLength(4);
        expect(onChangeMock).toHaveBeenCalledTimes(1);
        expect(onChangeMock).toHaveBeenCalledWith('');
    });
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Select value="test" onChange={() => null} />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - placeholder', () => {
        const wrapper = shallow(
            <Select value="test" onChange={() => null} disabled />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});
