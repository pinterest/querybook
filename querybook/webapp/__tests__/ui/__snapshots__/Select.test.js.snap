// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot - placeholder 1`] = `
<div
  className="control"
>
  <div
    className="Select"
  >
    <select
      disabled={true}
      onChange={[Function]}
      value="test"
    />
  </div>
</div>
`;

exports[`matches enzyme snapshots matches snapshot 1`] = `
<div
  className="control"
>
  <div
    className="Select"
  >
    <select
      disabled={false}
      onChange={[Function]}
      value="test"
    />
  </div>
</div>
`;
